app/controllers/concerns/foreman/controller/smart_proxy_auth.rb

```diff
@@ -81,9 +65,8 @@ def auth_smart_proxy(proxies = SmartProxy.unscoped.all, require_cert = true)
         logger.warn "No SSL cert supplied, falling back to reverse DNS for hostname lookup - request from #{request.remote_ip}"
         request_hosts = Resolv.new.getnames(request.remote_ip)
       end
-    elsif SETTINGS[:require_ssl]
-      logger.warn "SSL is required - request from #{request.remote_ip}"
     else
```

Comment:
Should we guard this and disallow this in production mode? This would make the `require_ssl` setting redundant here.
```suggestion
    elsif !Rails.env.production?
```

And I wonder if we should do the same for lines 64 - 67. Using reverse DNS is IMHO insecure. Perhaps also a separate PR that would get rid of the `require_ssl_smart_proxies` setting, but I really don't see the point in it. If you've set up HTTPS on Foreman, you can also set up HTTPS client certs IMHO.

Reply:
I agree with direction you're outlinng, but I'm against doing it in this PR or as part of this effort. As I said in the other PR, we seem to support pure HTTP in production for now. If we are dropping that (and I'd agree it is a good ide), we should do that explicitly and not make it a part of Rails upgrade. It's bloating the scope already. And I believe there should be RFC for that first (and I also believe we should provide another CA integration - or internal CA - having just the Puppet one, when we start forcing Users to have certificates is bad IMHO).

Reply:
I had the same thought yesterday. This and the related PR gave me the idea of writing up an RFC.

> and I also believe we should provide another CA integration - or internal CA - having just the Puppet one, when we start forcing Users to have certificates is bad IMHO

There is nothing that ties Foreman to the PuppetCA, but traditionally it has been a very convenient default.

Reply:
I opened https://community.theforeman.org/t/drop-require-ssl-and-require-ssl-smart-proxies-settings/26772 and https://github.com/theforeman/foreman/pull/9021.

Reply:
Could we merge this without the Production guard then and make the production guard in a followup?
I'll do that, but I'd rather not tie those up together in one PR.

Reply:
I'm not sure I'd feel comfortable with that as this PR is now. As it is now, I worry it could allow smart proxy authentication to happen over plain text HTTP. If you want to decouple it, I'd prefer to merge the production guard PR prior to merging this. Or you could leave the `elsif SETTINGS[:require_ssl]` in. That would keep it safe too.

Reply:
Oh. But if I read it correctly, it would never get to this branch unless the SSL is turned off and that was possible before this change as well :shrug: So this is not adding any issues, or is it? It just doesn't fix any.

Reply:
Technically there may be a very small chance, but what does removing the `elsif SETTINGS[:require_ssl]` branch here add? Keeping it gives certainty, removing it gives a chance of misusing the API to create insecure environments.

Reply:
Keeping it means keeping unreachable code :shrug:
But if you think it's better, I can keep that with comment.

Reply:
The alternative would be to let `require_ssl?` in `SmartProxyRequireSsl` always return `true`, but I think that would also be insufficient. The case I'd be worried about is (admittedly probably theoretical) where something has overridden `require_ssl?` to return `false` at the controller level. While we haven't upgraded to Rails 6.1 with a global rule for it, it can be reachable code. Keeping it in gives strong guarantees, which I generally speaking want when dealing with security.

Reply:
Well it can't hurt being explicit and I agree that if someone would succeed in hacking around that, we are in trouble :)
I've added the reachable branch (`!SETTINGS[:require_ssl]`) first and comment in the theoretically unreachable :)